presto-tpch/src/main/java/com/facebook/presto/tpch/TpchRecordSet.java

```diff
@@ -200,32 +208,35 @@ public void close()
 
         private boolean rowMatchesPredicate()
         {
-            if (!predicate.isPresent()) {
+            if (!predicate.isPresent() || predicate.get().isAll()) {
                 return true;
             }
-            return predicate.get().test(rowMap());
-        }
+            Set<ColumnHandle> predicatePushdownColumns = predicate.get().getDomains().get().keySet();
+            Map<ColumnHandle, NullableValue> rowMap = predicatePushdownColumns.stream()
+                    .collect(toImmutableMap(
+                            column -> column,
+                            column -> {
+                                TpchColumnHandle tpchColumnHandle = (TpchColumnHandle) column;
+                                Type type = tpchColumnHandle.getType();
+                                TpchColumn tpchColumn = table.getColumn(tpchColumnHandle.getColumnName());
+                                return NullableValue.of(type, getPrestoObject(tpchColumn, type));
+                            }));
 
-        private Map<ColumnHandle, NullableValue> rowMap()
-        {
-            ImmutableMap.Builder<ColumnHandle, NullableValue> builder = ImmutableMap.builder();
-            for (int field = 0; field < columnHandles.size(); ++field) {
-                Type type = columnTypes.get(field);
-                builder.put(columnHandles.get(field), NullableValue.of(type, getPrestoObject(field, type)));
-            }
-            return builder.build();
+            TupleDomain<ColumnHandle> rowTupleDomain = TupleDomain.fromFixedValues(rowMap);
+
+            return predicate.get().contains(rowTupleDomain);
```

Comment:
does it work when `predicate` does not contain all columns from row (I think in that case it will always return false)?

Reply:
Yes. Because `rowTupleDomain` contains only columns which are in predicate. It is build from `predicate` columns. See how `rowMap` is built.